packages/ec2-metadata-service/src/MetadataService.ts

```diff
@@ -0,0 +1,115 @@
+import { HttpRequest } from "@aws-sdk/protocol-http";
+import { HttpHandlerOptions } from "@aws-sdk/types";
+import { loadConfig } from "@smithy/node-config-provider";
+import { NodeHttpHandler } from "@smithy/node-http-handler";
+import { sdkStreamMixin } from "@smithy/util-stream";
+
+import { ENDPOINT_SELECTORS, IMDSv1_DISABLED_SELECTORS } from "./ConfigLoaders";
+import { MetadataServiceOptions } from "./MetadataServiceOptions";
+
+/**
+ * @public
+ */
+export class MetadataService {
+  private disableFetchToken: boolean;
+  private config: Promise<MetadataServiceOptions>;
+  /**
+   * Creates a new MetadataService object with a given set of options.
+   */
+  constructor(options: MetadataServiceOptions = {}) {
+    this.config = (async () => {
+      const profile = options?.profile || process.env.AWS_PROFILE;
+      return {
+        endpoint: options.endpoint ?? (await loadConfig(ENDPOINT_SELECTORS, { profile })()),
+        httpOptions: {
+          timeout: options?.httpOptions?.timeout || 0,
+        },
+        ec2MetadataV1Disabled:
+          options?.ec2MetadataV1Disabled ?? (await loadConfig(IMDSv1_DISABLED_SELECTORS, { profile })()),
+      };
+    })();
+    this.disableFetchToken = options?.disableFetchToken || false;
+  }
+
+  async request(path: string, options: { method?: string; headers?: Record<string, string> }): Promise<string> {
+    const { endpoint, ec2MetadataV1Disabled } = await this.config;
+    const handler = new NodeHttpHandler();
+    const endpointUrl = new URL(endpoint!);
+    const headers = options.headers || {};
+    /**
+     * If IMDSv1 is disabled and disableFetchToken is true, throw an error
+     * Refer: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/configuring-instance-metadata-service.html
+     */
+    if (this.disableFetchToken && ec2MetadataV1Disabled) {
+      throw new Error("IMDSv1 is disabled and fetching token is disabled, cannot make the request.");
+    }
+    /**
+     * Make request with token if disableFetchToken is not true (IMDSv2).
+     * Note that making the request call with token will result in an additional request to fetch the token.
+     */
+    if (!this.disableFetchToken) {
+      try {
+        headers["x-aws-ec2-metadata-token"] = await this.fetchMetadataToken();
+      } catch (err) {
+        if (ec2MetadataV1Disabled) {
+          // If IMDSv1 is disabled and token fetch fails (IMDSv2 fails), rethrow the error
+          throw err;
+        }
+        // If token fetch fails and IMDSv1 is not disabled, proceed without token (IMDSv1 fallback)
+      }
+    } // else, IMDSv1 fallback mode
+    const request = new HttpRequest({
+      method: options.method || "GET", // Default to GET if no method is specified
+      headers: headers,
+      hostname: endpointUrl.hostname,
+      path: endpointUrl.pathname + path,
+      protocol: endpointUrl.protocol,
+    });
```

Comment:
nit: newlines for readability
```suggestion

    const request = new HttpRequest({
      method: options.method || "GET", // Default to GET if no method is specified
      headers: headers,
      hostname: endpointUrl.hostname,
      path: endpointUrl.pathname + path,
      protocol: endpointUrl.protocol,
    });

```